File: leetcode/medium/all-nodes-distance-k-in-binary-tree/python/all-nodes-distance-k-in-binary-tree.py
# Solution for all-nodes-distance-k-in-binary-tree (Python)
# Problem: https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/
from collections import deque

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def distanceK(root, target, k):
    parent = {}
    def build_parent(node, par):
        if node:
            parent[node] = par
            build_parent(node.left, node)
            build_parent(node.right, node)
    build_parent(root, None)

    queue = deque([(target, 0)])
    visited = set([target])
    result = []

    while queue:
        node, distance = queue.popleft()

        if distance == k:
            result.append(node.val)
            continue # stop exploring this node
        # explore left child
        if node.left and node.left not in visited:
            visited.add(node.left)
            queue.append((node.left, distance + 1))
        # explore right child
        if node.right and node.right not in visited:
            visited.add(node.right)
            queue.append((node.right, distance + 1))
        # explore parent
        par = parent[node]
        if par and par not in visited:
            visited.add(par)
            queue.append((par, distance + 1))
    return result




if __name__ == '__main__':
    # Example tree construction
    #        3
    #       / \
    #      5   1
    #     / \ / \
    #    6  2 0  8
    #      / \
    #     7   4

    root = TreeNode(3)
    root.left = TreeNode(5)
    root.right = TreeNode(1)
    root.left.left = TreeNode(6)
    root.left.right = TreeNode(2)
    root.right.left = TreeNode(0)
    root.right.right = TreeNode(8)
    root.left.right.left = TreeNode(7)
    root.left.right.right = TreeNode(4)

    target = root.left  # Node with value 5
    k = 2
    print(distanceK(root, target, k))  # Output: [7, 4, 1]

File: leetcode/medium/all-nodes-distance-k-in-binary-tree/c/all-nodes-distance-k-in-binary-tree.c
// Solution for all-nodes-distance-k-in-binary-tree (C)
// Problem: https://leetcode.com/problems/all-nodes-distance-k-in-binary-tree/
#include <stdio.h>
#include <stblib.h>
#include <string.h>

// TreeNode Structure
typedef struct TreeNode{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;

// Queue structure for BFS
typedef struct QueueNode{
    TreeNode *treeNode;
    int distance;
    struct QueueNode *next;
} QueueNode;

typedef struct {
    QueueNode *front;
    QueueNode *rear;
} Queue;

//Function prototypes
void buildParentMap(TreeNode *node, TreeNode *parent, TreeNode **parentMap);
int* distancek(TreeNode *root, TreeNode *target, int k, int *return);
void enqueue(Queue *q, TreeNode *node, int distance);
QueueNode* dequeue(Queue *q);
int isEmpty(Queue *q);

//main solution function
int* distanceK(){


}

int main() {
    printf("Solution for all-nodes-distance-k-in-binary-tree\n");
    return 0;
}

